### PR TITLE
Rework prefetched data

### PR DESCRIPTION
**Pull request by lightsprint09**
Sorting and storing prefetched data should not be the responsibility of a ResurceDataProvider.

For sorting we should consider to provide a SortingArrayDataProvider which can wrap any ArrayDataProvider and sort it.

For prefetched data use `DataProviderSwitcher` introduced [here](https://github.com/lightsprint09/Sourcing/pull/26)

### CODE REVIEW

File: DBNetworkStack+Sourcing/ResourceDataProvider.swift [hunk 1]
```diff
@@ -75,15 +65,11 @@ public class ResourceDataProvider<Object>: ArrayDataProviding {
      - parameter mapFetchedObjectToArray: A function which maps a object to an array for using it as a dataSoruce. `nil` by default.
      - parameter delegate: A delegate for listing to events. `nil` by default.
      */
-    public init(resource: Resource<Array<Object>>?, prefetchedData: [Object] = [], networkService: NetworkServiceProviding,
+    public init(resource: Resource<Array<Object>>?, networkService: NetworkServiceProviding,
```
Comment (dennispost) [inline DBNetworkStack+Sourcing/ResourceDataProvider.swift]: Remove unnecessary parameters in function documentation
Add missing parameters

File: DBNetworkStack+Sourcing/ResourceDataProvider.swift [hunk 2]
```diff
@@ -164,11 +145,11 @@ public extension ResourceDataProvider {
      - parameter mapFetchedObjectToArray: A function which maps a object to an array for using it as a dataSoruce. `nil` by default.
      - parameter delegate: A delegate for listing to events. `nil` by default.
      */
-    public convenience init<R: ResourceModeling>(resource: R?, prefetchedData: [Object] = [],
-                            networkService: NetworkServiceProviding, whenStateChanges: @escaping ((ResourceDataProviderState) -> Void))
+    public convenience init<R: ResourceModeling>(resource: R?, networkService: NetworkServiceProviding,
+    whenStateChanges: @escaping ((ResourceDataProviderState) -> Void))
```
Comment (dennispost) [inline DBNetworkStack+Sourcing/ResourceDataProvider.swift]: Missing indent

File: DBNetworkStack+Sourcing/ResourceDataProvider.swift [hunk 3]
```diff
@@ -164,11 +145,11 @@ public extension ResourceDataProvider {
      - parameter mapFetchedObjectToArray: A function which maps a object to an array for using it as a dataSoruce. `nil` by default.
      - parameter delegate: A delegate for listing to events. `nil` by default.
      */
-    public convenience init<R: ResourceModeling>(resource: R?, prefetchedData: [Object] = [],
-                            networkService: NetworkServiceProviding, whenStateChanges: @escaping ((ResourceDataProviderState) -> Void))
+    public convenience init<R: ResourceModeling>(resource: R?, networkService: NetworkServiceProviding,
```
Comment (dennispost) [inline DBNetworkStack+Sourcing/ResourceDataProvider.swift]: Remove unnecessary parameters in function documentation
Add missing parameters

File: DBNetworkStack+SourcingTests/RessourceDataProviderTests.swift [hunk 4]
```diff
@@ -189,7 +144,7 @@ class ResourceDataProviderTests: XCTestCase {
     
     func testOnNetworkRequestCanceldWithNoEmptyData() {
```
Comment (dennispost) [inline DBNetworkStack+SourcingTests/RessourceDataProviderTests.swift]: Canceld -> Canceled

File: DBNetworkStack+SourcingTests/RessourceDataProviderTests.swift [hunk 5]
```diff
@@ -157,23 +129,6 @@ class ResourceDataProviderTests: XCTestCase {
...
-        dataProvider.sortDescriptor = { $0 < $1 }
-        dataProvider.load()
-        networkService.returnSuccess()
-        
-        //Then
-        let firstObject = dataProvider.object(at: IndexPath(item: 0, section: 0))
-        XCTAssertEqual(firstObject, 1)
-    }
-    
     func testOnNetworkRequestCanceldWithEmptyData() {
```
Comment (dennispost) [inline DBNetworkStack+SourcingTests/RessourceDataProviderTests.swift:134]: Canceld -> Canceled

**Review by ch-one — COMMENTED** [review]
Sorting data is essential, and i think it should be made as easy for a `ResourceDataProvider` as calling sort on an array.

Using `DataProviderSwitcher` to implement caching sounds good to me.

File: DBNetworkStack+Sourcing/ResourceDataProvider.swift [hunk 6]
```diff
@@ -122,12 +108,7 @@ public class ResourceDataProvider<Object>: ArrayDataProviding {
     
     func loadDidSucess(with newFetchedData: Array<Object>) {
```
Comment (dennispost) [inline DBNetworkStack+Sourcing/ResourceDataProvider.swift]: Cause there are no 'old fetched data' change 'newFetchedData' to just 'fetchedData'

**Review by dennispost — COMMENTED** [review]
- Check file structure esp. regarding 'MARK: private'
- Improve function documentation

**Comment by lightsprint09** [comment]
@ch-one the idea is to build a new SortedArrayDataProvider which can sort every array provider. So sorting could be shared between `ResourceDataProvider`, `ArrayDataProvider` and any future ArrayDataProviding.

**Review by ch-one — APPROVED** [review]

**Pull request merged**